pkg/scheduler/scheduler.go

```diff
@@ -280,7 +275,10 @@ func (s *Scheduler) FrontendLoop(frontend schedulerpb.SchedulerForFrontend_Front
 
 			enqueueSpan.Finish()
 		case schedulerpb.CANCEL:
-			s.cancelRequestAndRemoveFromPending(frontendAddress, msg.QueryID, "frontend cancelled query")
+			requestKey := queue.NewSchedulerRequestKey(frontendAddress, msg.QueryID)
+			schedulerReq := s.cancelRequestAndRemoveFromPending(requestKey, "frontend cancelled query")
+			// we may not have reached SubmitRequestSent for this query, but RequestQueue will handle this case
```

Comment:
Is there a risk of a race here, where we call `SubmitRequestCompleted` and then something else calls `SubmitRequestSent`? Would we still end up decrementing the component utilisation in that case?

Reply:
I don't believe we have a race:

case where the frontend cancel comes in before request was sent:
1. `SubmitRequestCompleted` results in a no-op here because it's not in the `RequestQueue`'s map
2. The request gets dequeued but never reaches the `forwardRequestToQuerier` step where it would be sent and incremented, because its context is already canceled, so it's never incremented.

case where frontend cancel comes in after the request was sent but before it was completed:
1. `SubmitRequestSent` called, incrementing
2. Cancel is handled, the request is canceled, and `SubmitRequestCompleted` is called here, decrementing.
3. The waiting `forwardRequestToQuerier` will get the propagated request cancellation, and exit, calling the deferred `SubmitRequestCompleted`, which no-ops because it was already called and removed from the map for this query.